SmartDeviceLink/SDLBackgroundTaskManager.m

```diff
@@ -41,26 +41,34 @@ - (void)startBackgroundTask {
 
     __weak typeof(self) weakSelf = self;
     self.currentBackgroundTaskId = [[UIApplication sharedApplication] beginBackgroundTaskWithName:self.backgroundTaskName expirationHandler:^{
+        // We have ~1 second to do cleanup before ending the background task. If we take too long, the system will kill the app.
         __strong typeof(weakSelf) strongSelf = weakSelf;
-        SDLLogD(@"The %@ background task expired", strongSelf.backgroundTaskName);
-        [strongSelf endBackgroundTask];
+        SDLLogD(@"The background task %@ expired", strongSelf.backgroundTaskName);
+        if (strongSelf.taskEndedHandler != nil) {
+            SDLLogD(@"Waiting for cleanup to finish before ending the background task");
+            strongSelf.taskEndedHandler();
```

Comment:
Doesn't this cause `endBackgroundTask` to never be set? It's not called from the `taskEndedHandler` in the `SecondaryTransportManager` which makes me think `endBackgroundTask` will ever be called in the `SecondaryTransportManager` if the task expires. If you were to require that the `endBackgroundTask` method be called from the handler, you should pass the `BackgroundTaskManager` as a parameter of the function.

Reply:
`endBackgroundTask` does get called in the `SDLSecondaryTransportManager`'s `disconnectSecondaryTransport` method. We need to wait for Core to ACK/NAK the end video and audio services and disconnect the TCP transport streams before ending the background task.

Reply:
If that's the case, then `disconnectSecondaryTransport` must be called every single time that the handler is called, right? And that definitely will not happen in the `if` and `else if` cases.

I'm also curious why the `else` goes to `configured` but the log says "Disconnecting TCP transport due to the background task ending." Configured doesn't disconnect the transport, it re-starts it, right? It should go to `stopped` or `reconnecting`? 

If any of my assumptions are wrong here, then the code is difficult to follow. I would request a refactor or clear comments describing what it supposed to be happening here.

Reply:
Just so we are on the same page does the following statement refer to the `if:else` in the expiration handler?  Or are you talking about the `if:else` in the `taskEndedHandler` block?
  
> If that's the case, then disconnectSecondaryTransport must be called every single time that the handler is called, right? And that definitely will not happen in the if and else if cases.

Reply:
Sorry, in `sdl_backgroundTaskEndedHandler`

Reply:
The issue should be fixed now.

Reply:
> I'm also curious why the else goes to configured but the log says "Disconnecting TCP transport due to the background task ending." Configured doesn't disconnect the transport, it re-starts it, right? It should go to stopped or reconnecting?

I updated the logs with more information. Basically it means that the manager is shutting down the secondary transport but it is still configured (i.e. has the ip and port) to restart the secondary transport.